option to hide clock for certain domains

chelovechishko opened the pull request:
...is here.
Also added auto-apply of changed settings on active tab. And css tricks to show this possibilities for all (probably).
(also have previous pr commit)

chelovechishko committed:
add settings for z-index

chelovechishko committed:
add domains list where clocks should not be shown

and all what it needed.
Also some rearrangements with html and css for more compact display
of previous fields and individual page for domains list.

chelovechishko committed:
auto apply changed settings in active tab

yaronf commented on options.css:
```diff
@@ -34,6 +59,58 @@ button {
+}
+
+/* Default values for opening settings page */
+#page_1_button {
```
The buttons now say "Page 1" and "Page 2". Maybe "General Settings" and "Excluded Domains"?

chelovechishko replied:
What about "domains", yes, they certainly now only for "not showing clocks there". But some may add some more additional buttons for them, like: dark/light theme for it (and not hide, of course) (not done, and my bet, not in near future).

yaronf replied:
Sounds good (capitalized: Domains).

But since the list is only **excluded** domains for now, please add explanatory text to say so.

yaronf replied:
Also, "auto apply" doesn't work for me.

chelovechishko replied:
You used "save button" for that? I transferred some changes from a pool of them, so i maybe missed some things.

yaronf replied:
I did. I'm getting an error (running on Chrome latest) for:

```
browser.runtime.onMessage.addListener( (request, sender, sendResponse) =>
```

Where `browser` should be `chrome`. Maybe it has something to do with it, although when I fixed it manually it still didn't work. 😞

chelovechishko replied:
Oh, yes, `browser` is a firefox thing, chrome only understands `chrome`.

yaronf commented on clock.js:
```diff
@@ -161,4 +165,27 @@ browser.runtime.onMessage.addListener( (request, sender, sendResponse) => {
+	}
 } );
+
+function check_is_page_in_domains_list( domains ) {
```
The domain detection thing is very confusing, because Chrome does not display the full name. E.g. when the URL box shows `ibm.com/uk-en` the real URL is `https://www.ibm.com/uk-en` (with a `www`). I spent some time not understanding the behavior before I realized that.

So instead of having the user enter domain names, maybe add a button "Exclude this domain"?

chelovechishko replied:
Yeah, chrome thingies, who would know, that they will be obstacles ¯\_(ツ)_/¯

> maybe add a button "Exclude this domain"?

Yes, would be conveniently, if somebody would add it.

yaronf commented on options.js:
```diff
@@ -41,20 +50,43 @@ function restore_options() {
+	switch_options_page( "page_2", "page_1" );
+	switch_options_page( "page_2_button", "page_1_button" );
+	} );
+document.getElementById( "switch_to_page_2" ).addEventListener( 'click', function() {
```
Page 2 only displays when the configuration is in full screen. In other words it doesn't show when I click the extension button.

![image](https://github.com/yaronf/overlay-clock/assets/1265636/72958421-827f-43da-965c-9e21867c22dd)

chelovechishko replied:
Yours extension is very narrow (while i have `min-page-width`). Dunno why. But i checked only in firefox with here and there changes, so who knows, does it affect it.

The strange thing is, i deleted `(0..1)` part in `background opacity`, but you still have it.

yaronf replied:
> Yours extension is very narrow (while i have `min-page-width`). Dunno why. But i checked only in firefox with here and there changes, so who knows, does it affect it.
> 

Any idea how to make the `browser`/`chrome` code work in both browsers?

> The strange thing is, i deleted `(0..1)` part in `background opacity`, but you still have it.

Yes, sorry. I have two Chrome instances (two "profiles") and I didn't even know that each one has its own extensions.

chelovechishko replied:
Oh, firefox is greater browser, so supports both ways.

chelovechishko committed:
changed simple number id names into something more elaborative

also add simple explanation of domains page

yaronf commented on options.js:
```diff
@@ -112,4 +140,146 @@ function send_message_to_tabs( tabs ) {
+	let domain_button = document.createElement( "input" );
+	domain_button.type = "button";
+	domain_button.title = "Delete this domain";
+	domain_button.value = "❌";
```
I'm not sure where the "+" is coming from (to add a domain) but please replace it with "➕".

chelovechishko replied:
> "➕"

Are you sure? While browsers have inner emoji fonts (or whatever has this symbol). My casual fonts don't have the symbol of yours, so i have it in text editor only as 2795 in square.

yaronf replied:
Yes, I'm sure. I see it correctly on FF and Chrome (Mac) as well as FF on Ubuntu. It's just a few characters after "❌" in the list of emojis.

chelovechishko replied:
That's why i said about casual text editor, not browsers. I'd prefer you commit such a change.

yaronf replied:
No problem. Where is it in the code? I couldn't find it.

chelovechishko replied:
`options.html` -> `id="add_domain"` -> `value=`

yaronf replied:
Committed into this PR (I used "vim").

yaronf commented:
I replaced all occurrences of "browser" to "chrome" in options.js, and tried in both browsers.

* Auto-apply doesn't work for me, testing with the color selectors.
* In Firefox: the color selectors only work in full-screen configuration. When I open the Settings as a dialog and press one of the colors, a color selector dialog pops up but the settings dialog closes!

chelovechishko commented:
> When I open the Settings as a dialog and press one of the colors, a color selector dialog pops up but the settings dialog closes!

Yeah, i am aware of this thing (but not committed here, i thought about do it in next pr). Firefox developers are also aware of it: 

```Text Color: <input title="In firefox you can only use this inside about:addons page as colorpicker window closes this popup page. See https://bugzilla.mozilla.org/show_bug.cgi?id=1378527"```

> Auto-apply doesn't work for me

As firefox has this bug, i tested it with change of z-index and through addition/deletion of opened domain. For me, it worked and with this branch.

chelovechishko commented:
Oh, yeah, forget to add: does add-on console output errors when you click on `save` button? `about:debugging#/runtime/this-firefox` -> `explore` (or something in english, dunno) button -> console.

chelovechishko committed:
make chrome more obedient

as it is not as good as Firefox.
Replaced all 'browser' occurences to 'chrome'.

yaronf commented:
> Oh, yeah, forget to add: does add-on console output errors when you click on `save` button? `about:debugging#/runtime/this-firefox` -> `explore` (or something in english, dunno) button -> console.

It's "inspect". But TBH I don't know how to open the extension (or the Settings dialog) in it, so I'm not able to see any errors.

chelovechishko commented:
![step 1](https://github.com/yaronf/overlay-clock/assets/14850730/88ccbdbf-e0b5-43ce-9cc9-338c7f7d9703)


![step 2](https://github.com/yaronf/overlay-clock/assets/14850730/bf7fdd7a-d513-4061-9112-f9277003fc0a)

yaronf commented:
When I open it **in Firefox** I get a `chrome:` URL at the top, instead of `moz-extension`! Which is probably why FF is confused and I'm not seeing the extension.

![image](https://github.com/yaronf/overlay-clock/assets/1265636/df11e478-ad52-4152-b931-447fb902050e)

chelovechishko commented:
Shit, if `options.js` has `chrome.tabs.query` instead of `browser.tabs.query` it just throws errors (in firefox). For some reason, firefox don't want `chrome` thing for this...

chelovechishko commented:
I have `chrome://` there too, but it does not matter for me.

yaronf committed:
Add "large plus" emoji

chelovechishko committed:
chrome: replace promises with callbacks

As chrome can't promises, even in manifest v3, resort to callbacks,
in dirty way.

chelovechishko committed:
make a variable for auto-updating changed settings strongly global

to not let erroring because of it (was previously just var in function)

chelovechishko commented:
Hope, with current changes, this will work even in chrome.

yaronf commented on options.js:
```diff
@@ -75,88 +75,101 @@ document.getElementById( "switch_to_page_domains_list" ).addEventListener( 'clic
+		try {
+			// console.log ("set_state_of_hide_checkbox: tab[0] = " + tab[0] +
+				// "; keys = " + Object.keys( tab[0] ) );
+			const sended_message = chrome.tabs.sendMessage(
```
```suggestion
			const sent_message = chrome.tabs.sendMessage(
```

yaronf committed:
typo

yaronf commented on options.js:
```diff
@@ -75,88 +75,101 @@ document.getElementById( "switch_to_page_domains_list" ).addEventListener( 'clic
+				tab[0].id,
+				{ visibility: "what" },
+				function ( response ) {
+					if( response.response == "hidden" ) {
```
Getting an error on Chrome:

```
Error handling response: TypeError: Cannot read properties of undefined (reading 'response')
```

chelovechishko replied:
Without arrow function, probably it will start to work (recent commit).

yaronf committed:
Lighter shade for options dialog for readability

yaronf commented:
@chelovechishko please fix the Domains page: the "+" is hard to read on its background and is a bit off, and we can accept this PR.

I will open a separate PR to bump the version number and (hopefully) move to Manifest V3, otherwise Chrome will soon drop this extension.

chelovechishko committed:
replace arrow function in onmessage.addlistener to casual one

Hope, now chrome could properly function in this listener.

chelovechishko committed:
return some colors to add_domain button

chelovechishko commented:
In my next pr i will add some comments about firefox vs color-selection thingy and such, and outline icon that in toolbar with white.

yaronf commented:
> In my next pr i will add some comments about firefox vs color-selection thingy and such, and outline icon that in toolbar with white.

The Settings are now nice and readable. One last thing, can you please align the "+" button with the "x" buttons:

![image](https://github.com/yaronf/overlay-clock/assets/1265636/9b07ef04-e19d-4593-935f-158c3ca39cc2)

yaronf commented:
> I will open a separate PR to bump the version number and (hopefully) move to Manifest V3, otherwise Chrome will soon drop this extension.

Will only bump the version number for now. The move to Manifest V3 is easy on Chrome but on Firefox it means the user needs to enable permissions on each tab separately. We will need to look into it.

chelovechishko commented:
> The Settings are now nice and readable

But domains are hanging in air, will turn on border.

chelovechishko committed:
force a + button to not lean on new line, but ignore it

Also turn on border-style for domains, and reduc inbetween space in half.
